notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+     "text": [
+      "                  week_start client_country       product event_type  quantity\n",
+      "0   2023-10-16T00:00:00.000Z          Japan  Magic 8-ball   forecast       949\n",
+      "1   2023-10-16T00:00:00.000Z          Japan  Magic 8-ball   purchase        15\n",
+      "2   2023-10-16T00:00:00.000Z          Japan      Pet rock   forecast      1004\n",
+      "3   2023-10-16T00:00:00.000Z          Japan      Pet rock   purchase        14\n",
+      "4   2023-10-16T00:00:00.000Z          Japan        Slinky   forecast       469\n",
+      "5   2023-10-16T00:00:00.000Z          Japan        Slinky   purchase        16\n",
+      "6   2023-10-23T00:00:00.000Z          Japan  Magic 8-ball   forecast       687\n",
+      "7   2023-10-23T00:00:00.000Z          Japan  Magic 8-ball   purchase         5\n",
+      "8   2023-10-23T00:00:00.000Z          Japan      Pet rock   forecast       745\n",
+      "9   2023-10-23T00:00:00.000Z          Japan      Pet rock   purchase        16\n",
+      "10  2023-10-23T00:00:00.000Z          Japan        Slinky   forecast       362\n",
+      "11  2023-10-23T00:00:00.000Z          Japan        Slinky   purchase        11\n",
+      "2023-10-26 01:11:57.615783 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 11, 57, 615783)"
+      ]
+     },
+     "execution_count": 32,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "sql = '''\n",
+    "    SELECT \n",
+    "       TIME_FLOOR(__time, 'P1W') week_start, \n",
+    "       \"client_country\",\n",
+    "       \"product\",\n",
+    "       \"event_type\",\n",
+    "       SUM(\"quantity\") \"quantity\"\n",
+    "    FROM \n",
+    "    (\n",
+    "        SELECT __time, \"client_country\", \"product\", \"event_type\", \"quantity\" FROM \"sales\"\n",
+    "    UNION ALL\n",
+    "        SELECT __time, \"client_country\", \"product\", \"event_type\", \"quantity\" FROM \"sales_forecast\"\n",
+    "    )\n",
+    "    WHERE  __time > TIME_FLOOR( CURRENT_TIMESTAMP, 'P1D') - INTERVAL '7' DAY\n",
+    "    AND \"product\" in ('Slinky', 'Magic 8-ball', 'Pet rock')\n",
+    "    AND \"client_country\" = 'Japan'\n",
+    "    GROUP BY 1,2,3,4\n",
+    "    ORDER BY 1,2,3,4\n",
+    "'''\n",
+    "\n",
+    "result=pandas.json_normalize(sql_client.sql(sql))\n",
+    "print_timestamp()\n",
+    "result"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e67a3e29-5822-4e1d-8c17-7d5382719d36",
+   "metadata": {},
+   "source": [
+    "Your specific results will vary because the data generator is somewhat random. In my results I can see that some country,product combinations do not have a forecast, so there is only a purchase row. An application processing these results would need to keep that in mind. The same would be true in sales if, for a particular period, there are no sales of a given product/country combo."
```

Comment:
Is there a way to make this consistent between runs? Concern is that some users may not see the technique being applied successfully.

Reply:
with static files, but the timestamps get old, so I'm leaning on generating on the fly. I think the scenario still plays out though, what didn't?